File: Machine Learning A-Z Template Folder/Part 1 - Data Preprocessing/Exercises/data_preprocessing.R
dataset = read.csv('Data.csv')
print(dataset)

#-------------------------------------------------------------------------------
# filling in missing data
#-------------------------------------------------------------------------------

dataset$Age = ifelse(is.na(dataset$Age),
                     ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Age)
dataset$Salary = ifelse(is.na(dataset$Salary),
                    ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),
                    dataset$Salary)
print(dataset)

#-------------------------------------------------------------------------------
# encoding categorical data
#-------------------------------------------------------------------------------

dataset$Country = factor(dataset$Country,
                         levels = c('France', 'Spain', 'Germany'),
                         labels = c(1, 2, 3))
dataset$Purchased = factor(dataset$Purchased,
                        levels = c('No', 'Yes'),
                        labels = c(0, 1))
print(dataset)

#-------------------------------------------------------------------------------
# bringing in library for train test split
#-------------------------------------------------------------------------------

library(caTools)
set.seed(123)

split = sample.split(dataset$Purchased, SplitRatio = 0.8) # SplitRatio is for train_size
print(split)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)
print(training_set)
print(test_set)

#-------------------------------------------------------------------------------
# Feature Scaling
#-------------------------------------------------------------------------------

training_set[, 2:3] = scale(training_set[, 2:3])
test_set[, 2:3] = scale(test_set[, 2:3])
print(training_set)
print(test_set)

File: Machine Learning A-Z Template Folder/Part 1 - Data Preprocessing/Exercises/data_preprocessing.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import Imputer, LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split


#-------------------------------------------------------------------------------
#                            Data Preprocessing
#-------------------------------------------------------------------------------

dataset = pd.read_csv('Data.csv')
x = dataset.iloc[:, :-1].values # all rows, all columns except the last
y = dataset.iloc[:, 3].values # all rows, only last column
print('\n-------------------------------------------------------------------\n')
print(x)
print('\n-------------------------------------------------------------------\n')
print(y)
print('\n-------------------------------------------------------------------\n')

#-------------------------------------------------------------------------------
#                         Filling in Missing Data
#-------------------------------------------------------------------------------

imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
imputer_x = imputer.fit(x[:, 1:3])
x[:, 1:3] = imputer_x.transform(x[:, 1:3])
print(x)
print('\n-------------------------------------------------------------------\n')

#-------------------------------------------------------------------------------
#                        Encoding Categorical Data
#-------------------------------------------------------------------------------

encoder_x = LabelEncoder()
x[:, 0] = encoder_x.fit_transform(x[:, 0])
print(x)
print('\n-------------------------------------------------------------------\n')

hotEncoder_x = OneHotEncoder(categorical_features = [0])
x = hotEncoder_x.fit_transform(x).toarray()
print(x)
print('\n-------------------------------------------------------------------\n')

encoder_y = LabelEncoder()
y = encoder_y.fit_transform(y)
print(y)
print('\n-------------------------------------------------------------------\n')

#-------------------------------------------------------------------------------
#                        train test split of data
#-------------------------------------------------------------------------------

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 0)

#-------------------------------------------------------------------------------
#                        Feature Scaling
#-------------------------------------------------------------------------------

scaler_x = StandardScaler()
scaler_y = StandardScaler()

X_train = scaler_x.fit_transform(X_train)
X_test = scaler_x.transform(X_test)
